sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UseUriInsteadOfString.cs

```diff
@@ -51,75 +53,143 @@ public sealed class UseUriInsteadOfString : SonarDiagnosticAnalyzer
+                        !propertySymbol.IsOverride &&
+                        NameContainsUri(propertySymbol.Name))
                     {
-                        c.ReportDiagnostic(Diagnostic.Create(rule_S3995, methodDeclaration.ReturnType.GetLocation()));
+                        c.ReportDiagnostic(Diagnostic.Create(rule_S3996, propertyDeclaration.Type.GetLocation()));
                     }
+                },
+                SyntaxKind.PropertyDeclaration);
 
-                    var stringUrlParams = GetStringUrlParamIndexes(methodSymbol);
-                    if (!stringUrlParams.Any() ||
-                        HasOverloadThatUsesUriTypeInPlaceOfString(methodSymbol, stringUrlParams))
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var invokedMethodSymbol = c.SemanticModel.GetSymbolInfo(c.Node).Symbol as IMethodSymbol;
+                    if (invokedMethodSymbol == null || invokedMethodSymbol.IsInType(KnownType.System_Uri))
                     {
                         return;
                     }
 
-                    foreach (int paramIdx in stringUrlParams)
+                    var stringUrlParams = GetStringUrlParamIndexes(invokedMethodSymbol);
+                    var methodOverloads = FindOverloadsThatUseUriTypeInPlaceOfString(invokedMethodSymbol, stringUrlParams);
+                    if (stringUrlParams.Count > 0 && methodOverloads.Any())
                     {
-                        c.ReportDiagnostic(Diagnostic.Create(rule_S3994,
-                            methodDeclaration.ParameterList.Parameters[paramIdx].Type.GetLocation()));
+                        c.ReportDiagnostic(Diagnostic.Create(rule_S4005, c.Node.GetLocation()));
                     }
                 },
-                SyntaxKind.MethodDeclaration);
+                SyntaxKind.InvocationExpression,
+                SyntaxKind.ObjectCreationExpression);
+        }
 
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
+        private void VerifyDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            var methodDeclaration = (BaseMethodDeclarationSyntax)context.Node;
+            var methodSymbol = context.SemanticModel.GetDeclaredSymbol(methodDeclaration);
+
+            if (methodDeclaration == null || methodSymbol == null || methodSymbol.IsOverride)
+            {
+                return;
+            }
+
+            VerifyReturnType(context, methodDeclaration, methodSymbol);
+
+            var stringUrlParams = GetStringUrlParamIndexes(methodSymbol);
+            if (!stringUrlParams.Any())
+            {
+                return;
+            }
+
+            var methodOverloads = FindOverloadsThatUseUriTypeInPlaceOfString(methodSymbol, stringUrlParams);
+            if (methodOverloads.Any())
+            {
+                var methodOverloadSet = new HashSet<IMethodSymbol>(methodOverloads);
```

Comment:
That would be true if the set was always created (that is, if method always had an overload). If there are no overloads, we can avoid creating a set, which saves time.
I've written a quick test and it shows the speed is comparable if 90% of methods have overloads. Any less than that and version with Any() should be quicker.